src/context/defrag.rs

```diff
@@ -0,0 +1,226 @@
+                self.defrag_ctx,
+                cursor,
+            )
+        }
+        .into()
+    }
+
+    /// Fetch a cursor value that has been previously stored using [`Self::set_cursor`].
+    /// If not called for a late defrag operation, [`Err`] will be returned.
+    pub fn get_cursor(&self) -> Result<u64, RedisError> {
+        let mut cursor: u64 = 0;
+        let res: Status = unsafe {
+            RedisModule_DefragCursorGet.expect("RedisModule_DefragCursorGet is NULL")(
+                self.defrag_ctx,
+                (&mut cursor) as *mut u64,
+            )
+        }
+        .into();
+        if res == Status::Ok {
+            Ok(cursor)
+        } else {
+            Err(RedisError::Str("Could not get cursor value"))
+        }
+    }
+
+    /// Defrag a memory allocation previously allocated by RM_Alloc, RM_Calloc, etc.
+    /// The defragmentation process involves allocating a new memory block and copying
+    /// the contents to it, like realloc().
+    ///
+    /// If defragmentation was not necessary, NULL is returned and the operation has
+    /// no other effect.
+    ///
+    /// If a non-NULL value is returned, the caller should use the new pointer instead
+    /// of the old one and update any reference to the old pointer, which must not
+    /// be used again.
+    ///
+    /// The function is unsafe because it is assumed that the pointer is valid and previusly
+    /// allocated. It is considered undefined if this is not the case.
+    pub unsafe fn defrag_realloc<T>(&self, mut ptr: *mut T) -> *mut T {
+        let new_ptr: *mut T = RedisModule_DefragAlloc.expect("RedisModule_DefragAlloc is NULL")(
+            self.defrag_ctx,
+            ptr.cast(),
+        )
+        .cast();
+        if !new_ptr.is_null() {
+            ptr = new_ptr;
+        }
+        ptr
+    }
+
+    /// Allocate memory using defrag allocator if supported by the
+    /// current Redis server, fallback to regular allocation otherwise.
+    pub fn defrag_alloc<T>(&self, layout: Layout) -> *mut T {
+        unsafe { std::alloc::alloc(layout) }.cast()
+    }
+
+    /// Deallocate memory using defrag deallocator if supported by the
+    /// current Redis server, fallback to regular deallocation otherwise.
+    pub fn defrag_dealloc<T>(&self, ptr: *mut T, layout: Layout) {
+        unsafe { std::alloc::dealloc(ptr.cast(), layout) }
+    }
+
+    /// Defrag a [RedisString]
+    ///
+    /// NOTE: It is only possible to defrag strings that have a single reference.
+    /// Typically this means strings that was copy/cloned using [RedisString::safe_clone]
+    /// or created using [RedisString::new] will not be defrag and will be returned as is.
+    pub fn defrag_redis_string(&self, mut s: RedisString) -> RedisString {
+        let new_inner = unsafe {
+            RedisModule_DefragRedisModuleString
+                .expect("RedisModule_DefragRedisModuleString is NULL")(
```

Comment:
`RedisModule_DefragRedisModuleString.expect("RedisModule_DefragRedisModuleString is available")`